Passing to C issue

### Describe the bug

Here https://github.com/antono2/tmp/blob/603637399d22d4205af468cd6b634f2f64eefdb1/main.v#L118 I'd like to pass a pointer to C, but get 
```
error: invalid initializer
 6334 |         main__Struct_array __tmp_arg_1 = param;
```

And using `C.struct_array` instead of alias
```
error: unknown type name ‘C__struct_array’; did you mean ‘struct_array’?
 6334 |         C__struct_array __tmp_arg_1 = param;
```

### Reproduction Steps

git clone --depth=1 https://github.com/antono2/tmp && cd tmp && git checkout 603637399d22d4205af468cd6b634f2f64eefdb1 && v -show-c-output -cc gcc run .


### Expected Behavior

Call set_struct_array in include/array.c

### Current Behavior

```
/tmp/v_1000/tmp.01JPPP02KCQ7864DRP1NQT4M0Q.tmp.c: In function ‘main__set_struct_array’:
/tmp/v_1000/tmp.01JPPP02KCQ7864DRP1NQT4M0Q.tmp.c:6334:42: error: invalid initializer
 6334 |         main__Struct_array __tmp_arg_1 = param;
      |                                          ^~~~~
/tmp/v_1000/tmp.01JPPP02KCQ7864DRP1NQT4M0Q.tmp.c:6335:26: error: incompatible type for argument 1 of ‘set_struct_array’
 6335 |         set_struct_array(__tmp_arg_1);
      |                          ^~~~~~~~~~~
      |                          |
      |                          main__Struct_array {aka struct_array}
/home/anton/workspace/tmp/include/array.c:132:32: note: expected ‘struct_array *’ but argument is of type ‘main__Struct_array’ {aka ‘struct_array’}
  132 | set_struct_array(struct_array* param)
      |                  ~~~~~~~~~~~~~~^~~~~

```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.9 b528395

### Environment details (OS name and version, etc.)

|V full version      |V 0.4.9 9921692.b528395
|:-------------------|:-------------------
|OS                  |linux, Ubuntu 24.04.2 LTS
|Processor           |8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz
|Memory              |0.44GB/15.56GB
|                    |
|V executable        |/home/anton/workspace/v/v
|V last modified time|2025-03-19 02:20:05
|                    |
|V home dir          |OK, value: /home/anton/workspace/v
|VMODULES            |OK, value: /home/anton/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/anton
|                    |
|Git version         |git version 2.43.0
|V git status        |weekly.2025.08-198-gb5283950
|.git/config present |true
|                    |
|cc version          |cc (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
|gcc version         |gcc (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
|clang version       |clang version 21.0.0git (https://github.com/llvm/llvm-project.git bd034ab1113cd3705fedf75d870dc4b7388d8f70)
|tcc version         |tcc version 0.9.28rc 2024-07-31 HEAD@1cee0908 (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 0134e9b9-dirty
|emcc version        |N/A
|glibc version       |ldd (Ubuntu GLIBC 2.39-0ubuntu8.4) 2.39

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -2577,7 +2577,7 @@ fn (mut g Gen) keep_alive_call_pregen(node ast.CallExpr) int {
 		// evaluation order is preserved
 		expected_type := node.expected_arg_types[i]
 		typ_sym := g.table.sym(expected_type)
-		typ := typ_sym.cname
+		typ := g.styp(expected_type)
 		if typ_sym.kind != .array_fixed {
 			g.write('${typ} __tmp_arg_${tmp_cnt_save + i} = ')
 			g.ref_or_deref_arg(arg, expected_type, node.language, false)

--- a/vlib/v/tests/c_function/array.c
+++ b/vlib/v/tests/c_function/array.c
@@ -1,0 +1,135 @@
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+
+typedef struct {
+  char **array_ptr;
+  int array_len;
+  int string_len;
+} array_t;
+
+typedef enum flag_bits {
+	AAA = 0,
+	BBB = 1,
+	CCC = 2
+} flag_bits;
+
+typedef enum flag_bits_main {
+	AAAA = 0,
+	BBBB = 1,
+	CCCC = 2
+} flag_bits_main;
+
+typedef struct {
+  array_t arr[4];
+  flag_bits enu[2];
+} struct_array;
+typedef uint32_t flags;
+
+typedef union union_t {
+    float       float32[4];
+    int32_t     int32[4];
+    uint32_t    uint32[4];
+} union_t;
+
+array_t*
+array_string_new(int array_len, int string_len)
+{
+  int i;
+  char **array_ptr = (char**) malloc(array_len * sizeof(char**));
+
+  for(i = 0; i < array_len; i++) {
+    array_ptr[i] = (char*) malloc(string_len * sizeof(char));
+  }
+
+  array_t *array = (array_t*) malloc(sizeof(array_t*));
+  array->array_ptr = array_ptr;
+  array->array_len = array_len;
+  array->string_len = string_len;
+
+  return array;
+}
+
+int
+array_string_set(array_t *array, int index, char *string)
+{
+  strncpy(array->array_ptr[index], string, array->string_len);
+  return 0;
+}
+
+char*
+array_string_get(array_t *array, int index)
+{
+  return array->array_ptr[index];
+}
+
+int
+array_string_len(array_t *array)
+{
+  return array->array_len;
+}
+
+int
+array_string_free(array_t *array)
+{
+  int i;
+  for(i = 0; i < array->array_len; i++) {
+    free(array->array_ptr[i]);
+  }
+  free(array->array_ptr);
+  return 0;
+}
+
+char** get_string_array() {
+  array_t *array = array_string_new(4, 4);
+
+  array_string_set(array, 0, "foo");
+  array_string_set(array, 1, "bar");
+  array_string_set(array, 2, "bat");
+  array_string_set(array, 3, ".... overflowed string");
+
+	return array->array_ptr;
+}
+
+struct_array get_struct_array() {
+  int i;
+  array_t* array = array_string_new(4, 4);
+
+  array_string_set(array, 0, "foo");
+  array_string_set(array, 1, "bar");
+  array_string_set(array, 2, "bat");
+  array_string_set(array, 3, ".... overflowed string");
+
+  struct_array ret;
+  for(i = 0; i < array_string_len(array); i++) {
+    ret.arr[i] = *array;
+  };
+  
+  return ret;
+}
+
+flags get_enum() {
+  flag_bits ret = AAA;
+  return ret;
+}
+
+union_t get_union() {
+  uint32_t r = 123;
+	union_t ret;
+	ret.uint32[0] = r;
+	ret.uint32[1] = r;
+	ret.uint32[2] = r;
+	ret.uint32[3] = r;
+	return ret;
+}
+
+void get_stub(flag_bits param[2]) {}
+
+void get_stub_main(flag_bits_main param[2]) {}
+
+void
+set_struct_array(struct_array* param)
+{
+  puts("\nset struct_array\n");
+}

--- a/vlib/v/tests/c_function/array.c
+++ b/vlib/v/tests/c_function/array.c
@@ -43,7 +43,7 @@ array_string_new(int array_len, int string_len)
     array_ptr[i] = (char*) malloc(string_len * sizeof(char));
   }
 
-  array_t *array = (array_t*) malloc(sizeof(array_t*));
+  array_t *array = (array_t*) malloc(sizeof(array_t));
   array->array_ptr = array_ptr;
   array->array_len = array_len;
   array->string_len = string_len;
@@ -105,6 +105,7 @@ struct_array get_struct_array() {
   for(i = 0; i < array_string_len(array); i++) {
     ret.arr[i] = *array;
   };
+  free(array);
   
   return ret;
 }

--- a/vlib/v/tests/c_function/pass_ref_test.c.v
+++ b/vlib/v/tests/c_function/pass_ref_test.c.v
@@ -1,0 +1,53 @@
+module main
+
+#include "@VMODROOT/array.c"
+
+enum Flag_bits {
+	aaa = 0
+	bbb = 1
+	ccc = 2
+}
+
+pub type Array_t = C.array_t
+
+@[typedef]
+struct C.array_t {
+pub mut:
+	array_ptr  &&char
+	array_len  int
+	string_len int
+}
+
+pub type Struct_array = C.struct_array
+
+@[typedef]
+struct C.struct_array {
+pub mut:
+	arr [4]Array_t
+	enu [2]Flag_bits
+}
+
+fn C.get_string_array() &&char
+@[inline]
+pub fn get_string_array() &&char {
+	return C.get_string_array()
+}
+
+fn C.get_struct_array() Struct_array
+@[inline]
+pub fn get_struct_array() Struct_array {
+	return C.get_struct_array()
+}
+
+@[keep_args_alive]
+fn C.set_struct_array(&Struct_array)
+@[inline]
+pub fn set_struct_array(param &Struct_array) {
+	C.set_struct_array(param)
+}
+
+fn test_main() {
+	struct_array := get_struct_array()
+	set_struct_array(&Struct_array{})
+	set_struct_array(&struct_array)
+}

--- a/vlib/v/tests/c_function/pass_ref_test.c.v
+++ b/vlib/v/tests/c_function/pass_ref_test.c.v
@@ -39,6 +39,8 @@ pub fn get_struct_array() Struct_array {
 	return C.get_struct_array()
 }
 
+fn C.array_string_free(&Array_t)
+
 @[keep_args_alive]
 fn C.set_struct_array(&Struct_array)
 @[inline]
@@ -48,6 +50,9 @@ pub fn set_struct_array(param &Struct_array) {
 
 fn test_main() {
 	struct_array := get_struct_array()
+
 	set_struct_array(&Struct_array{})
 	set_struct_array(&struct_array)
+
+	C.array_string_free(&struct_array.arr[0])
 }

